### ISSUE TITLE
.limit().all() and .offset().all() seem to override .options(load_only('col_1', 'col_2')) and include all columns

### ISSUE THREAD

**Issue by christopherpickering** [sqlalchemy/sqlalchemy#5457]
**Describe the bug**

Hi, I'm trying to use "load_only" to select only a handful of columns in the table, however, it seems to always return all columns in the table, not just what I specify.

```python
stuff = Table.query.filter_by(col_1='this')).options(load_only('col_1', 'col_2')).limit(5).all()
```

stuff should return data with only two columns, col 1 and 2, but is returning all columns in the table.

so if I put just .limit(5), I get the sql query that is limited to the one column. When I add .all() to get the results then I get all columns. With offset it is the same result - nice sql query until I add ".all()" to get the results.

**Expected behavior**
I expect to only have the two columns in the result when I "load_only".

**Versions.**
 - OS: Alpine Linux
 - Python:3.8.2
 - SQLAlchemy:1.3.17
 - Database:Sqlite3
 - DBAPI:

**Have a nice day!**

ref: https://github.com/pallets/flask-sqlalchemy/issues/851

**Comment by zzzeek** [comment]
hi -

can you please provide a minimal, reproducing model for Thing and also illustrate the SQL output for "echo='debug'" when you run this query, thanks.

**Comment by zzzeek** [comment]
I have a feeling you are actually looking for:

    session.query(Thing.col1, Thing.col2)

if you want tuples with just those two columns.  load_only() doesnt' prevent you from getting Thing objects back.

**Comment by christopherpickering** [comment]
@zzzeek thanks for the comment, I appreciate it!

Here is a full example code -
```python
import sqlalchemy
from sqlalchemy.orm import load_only
from sqlalchemy import Column,Integer,String
from sqlalchemy.ext.declarative import declarative_base
from dataclasses import dataclass
from sqlalchemy import create_engine
engine = create_engine('sqlite:///db.sqlite', echo = 'debug')
Base = declarative_base()

@dataclass
class Job(Base):
  id: int
  name: str
  description: str
  enabled: int
  __tablename__ = 'job'

  id = Column(Integer, primary_key=True, index=True)
  name = Column(String(120), nullable=True)
  description = Column(String(8000), nullable=False)
  enabled = Column(Integer, nullable=True)


from sqlalchemy.orm import sessionmaker
session = sessionmaker()
session.configure(bind=engine)
Base.metadata.create_all(engine)


j = Job(name='Christopher',description='fun stuff',enabled=1)
s=session()
s.add(j)
s.commit()

print(s.query(Job).options(load_only('name'))) #.all()

print(s.query(Job).options(load_only('name')).all())
```

and the output -
```sh
python test.py
2020-07-14 07:42:10,848 INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS VARCHAR(60)) AS anon_1
2020-07-14 07:42:10,852 INFO sqlalchemy.engine.base.Engine ()
2020-07-14 07:42:10,855 INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' 
AS VARCHAR(60)) AS anon_1
2020-07-14 07:42:10,863 INFO sqlalchemy.engine.base.Engine ()
2020-07-14 07:42:10,866 INFO sqlalchemy.engine.base.Engine PRAGMA main.table_info("job")      
2020-07-14 07:42:10,867 INFO sqlalchemy.engine.base.Engine ()
2020-07-14 07:42:10,869 DEBUG sqlalchemy.engine.base.Engine Col ('cid', 'name', 'type', 'notnull', 'dflt_value', 'pk')
2020-07-14 07:42:10,870 INFO sqlalchemy.engine.base.Engine PRAGMA temp.table_info("job")      
2020-07-14 07:42:10,872 INFO sqlalchemy.engine.base.Engine ()
2020-07-14 07:42:10,872 DEBUG sqlalchemy.engine.base.Engine Col ('cid', 'name', 'type', 'notnull', 'dflt_value', 'pk')
2020-07-14 07:42:10,880 INFO sqlalchemy.engine.base.Engine
CREATE TABLE job (
        id INTEGER NOT NULL,
        name VARCHAR(120),
        description VARCHAR(8000) NOT NULL,
        enabled INTEGER,
        PRIMARY KEY (id)
)


2020-07-14 07:42:10,890 INFO sqlalchemy.engine.base.Engine ()
2020-07-14 07:42:10,894 INFO sqlalchemy.engine.base.Engine COMMIT
2020-07-14 07:42:10,896 INFO sqlalchemy.engine.base.Engine CREATE INDEX ix_job_id ON job (id)
2020-07-14 07:42:10,897 INFO sqlalchemy.engine.base.Engine ()
2020-07-14 07:42:10,899 INFO sqlalchemy.engine.base.Engine COMMIT
2020-07-14 07:42:10,903 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2020-07-14 07:42:10,906 INFO sqlalchemy.engine.base.Engine INSERT INTO job (name, description, enabled) VALUES (?, ?, ?)
2020-07-14 07:42:10,909 INFO sqlalchemy.engine.base.Engine ('Christopher', 'fun stuff', 1)    
2020-07-14 07:42:10,912 INFO sqlalchemy.engine.base.Engine COMMIT
SELECT job.id AS job_id, job.name AS job_name 
FROM job
2020-07-14 07:42:10,918 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2020-07-14 07:42:10,919 INFO sqlalchemy.engine.base.Engine SELECT job.id AS job_id, job.name AS job_name
FROM job
2020-07-14 07:42:10,922 INFO sqlalchemy.engine.base.Engine ()
2020-07-14 07:42:10,923 DEBUG sqlalchemy.engine.base.Engine Col ('job_id', 'job_name')
2020-07-14 07:42:10,925 DEBUG sqlalchemy.engine.base.Engine Row (1, 'Christopher')
2020-07-14 07:42:10,927 INFO sqlalchemy.engine.base.Engine SELECT job.description AS job_description, job.enabled AS job_enabled
FROM job
WHERE job.id = ?
2020-07-14 07:42:10,930 INFO sqlalchemy.engine.base.Engine (1,)
2020-07-14 07:42:10,932 DEBUG sqlalchemy.engine.base.Engine Col ('job_description', 'job_enabled')
2020-07-14 07:42:10,934 DEBUG sqlalchemy.engine.base.Engine Row ('fun stuff', 1)
[Job(id=1, name='Christopher', description='fun stuff', enabled=1)]
```

In this example the sql statement coming from line 35 is selecting description and enabled fields, even though I have just name to be selected.. I'm not sure why, when I initially created the issue I was getting just the column I requested :) 

What do you think? Thanks!

**Comment by christopherpickering** [comment]
> I have a feeling you are actually looking for:
> 
> ```
> session.query(Thing.col1, Thing.col2)
> ```
> 
> if you want tuples with just those two columns. load_only() doesnt' prevent you from getting Thing objects back.

This worked well, thanks!

**Comment by zzzeek** [comment]
it looks like your Job class, when you call __repr__ on it, it accessing all the fields, that is, including "description" and "enabled".  I would assume this is some behavior of "dataclasses".   SQLAlchemy ORM objects emit a deferred load for additional columns when you request them.   

your example is running the primary SELECT only once when you say .all().  The previous print() you have there is not running the query, only printing it.  you need to iterate the query to get results or otherwise call methods like .all().

So in your output we have, your print statement:

```
SELECT job.id AS job_id, job.name AS job_name 
FROM job

```

then your .all() call:

```
2020-07-14 07:42:10,919 INFO sqlalchemy.engine.base.Engine SELECT job.id AS job_id, job.name AS job_name
FROM job
2020-07-14 07:42:10,922 INFO sqlalchemy.engine.base.Engine ()
2020-07-14 07:42:10,923 DEBUG sqlalchemy.engine.base.Engine Col ('job_id', 'job_name')
2020-07-14 07:42:10,925 DEBUG sqlalchemy.engine.base.Engine Row (1, 'Christopher')

```


then when you print the object, whatever is doing `__str__` or `__repr__` is requesting the other attributes, so SQLAlchemy loads these:

```
2020-07-14 07:42:10,927 INFO sqlalchemy.engine.base.Engine SELECT job.description AS job_description, job.enabled AS job_enabled
FROM job
WHERE job.id = ?
2020-07-14 07:42:10,930 INFO sqlalchemy.engine.base.Engine (1,)
2020-07-14 07:42:10,932 DEBUG sqlalchemy.engine.base.Engine Col ('job_description', 'job_enabled')

```

background on how all this works is at https://docs.sqlalchemy.org/en/13/orm/loading_columns.html#deferred-column-loading .

SQLAlchemy 1.4 will also include a "raiseload" option that would force it to raise an exception when those extra columns are loaded but it's clear here you just need tuples.
